### DISCUSSION TITLE
Dexie Cloud behavior: Operations are undone when syncing

### DISCUSSION THREAD

**Discussion by dfahlander** [dexie/Dexie.js#2024] [General]
In Dexie Cloud, if the server need to reject a change, it will send updates to the client that will reset it to the same state as the server - without the requested changes - it could be added objects, updated properties or deleted objects. The reason for this is a principle that Dexie Cloud always keep: After sync, the partial database of the client will mirror exactly the corresponding part on the server. Logically, if the client changes was rejected by the server, it needs to reset the client as if the changes were never done.

If you ever experience this behavior, ***make sure to open Devtools and look in the Network tab and inspect the content of the sync response***. You will find a property "rejections".

![image](https://github.com/dexie/Dexie.js/assets/6791329/f7227ca3-faf1-4769-88a5-e2ab078560fa)

It's an array of exceptions and the corresponding transaction ID for any of the reasons mentioned below:

* Permission was denied
* A constraint error happened (like duplicate key on add operation)
* Another exception happened that made Dexie Cloud fail to store the changes

If you have any questions regarding this behavior, you are welcome to discuss it here.

**Comment by gibbage** [comment]
Thanks for the tip @dfahlander. I was able to use this method to see the rejection message. Unfortunately, I'm still not quite sure how to solve the problem. I have a feeling it's something to do with not managing my schema versions and migrations correctly.

I have created a "config" table, that I want to store app configuration in. It was simply meant to be a key/value store, e.g.

```ts
type Config = {
	key: string;
	value: string;
};
```

I added it to the latest version of the schema. I didn't create a version 3, which I think I should have:

```ts
export class MyDexieDb extends Dexie {
	// other tables
	// ...
	config: DexieCloudTable<Config, 'key'>;

	constructor() {
		const addons = [dexieCloud];
		super('MyDexieDb', { addons });
		this.version(1).stores({
			 otherTable: `@id`,
			// ...
		});
		this.version(2).stores({
			anotherTableCreatedEarlier: `@id`,
			config: `key`
		});
	}
}
```

I had a function that I ran to populate the various tables whenever the user was logged in and syncing complete. Within that function I was checking for entries in the `config` table. If it was empty I was adding a key/value. I'd see it flash on the screen and then be removed after a push. This issue was only happening for a particular key. I can actually add more records to the config table, but if I add a record with that particular primary key, the server rejects it. Using the method you described above I get an error that makes sense for what I'm seeing.

I see the `upsert` the `changes` array on the request body JSON. Then in the response, I see the `rejection` and subsequent `delete` citing that `Object { name: "error", message: 'duplicate key value violates unique constraint "objects_pkey"', txid: "kF...A==" }`. But when I do a `db.config.toArray()` there is not a row with that key as the primary key.

![dexie-config-request](https://github.com/dexie/Dexie.js/assets/832653/fe16ace9-1357-4722-86e9-b77bd0d13789)
![dexie-config-response](https://github.com/dexie/Dexie.js/assets/832653/5294846a-9da0-40e3-9707-f4a4f5ce54b9)

I've also exported the data using the CLI tool using `npx dexie-cloud export export.json`. When I look at the data, there is no duplicate key in the `config` table for the broken dev test account.

This is only happening for one of my dev test accounts. I have a prod account that's currently working and has the same key/value in the `config` table. Any ideas what's going on here?
